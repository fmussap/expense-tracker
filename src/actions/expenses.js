import database from '../firebase/firebase'
import * as actions from './actions-type'

export const createExpense = (expense) => {
  return ({
    type: actions.ADD_EXPENSE,
    expense
  })
}

export const addExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }
    return database.ref(`users/${uid}/expenses`).push(expense)
      .then((ref) => {
        dispatch(createExpense({
          id: ref.key,
          ...expense
        }))
      })
      .catch((e) => {
        console.log('Error on add expense', e)
      })
  }
}

export const removeExpenseState = ({ id } = {}) => ({
  type: actions.REMOVE_EXPENSE,
  id
})

export const removeExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove()
      .then(() => {
        dispatch(removeExpenseState({ id }))
      })
      .catch((e) => {
        console.log('Error on remove expense', e)
      })
  }
}

export const editExpenseState = (id, updates) => ({
  type: actions.EDIT_EXPENSE,
  id,
  updates
})

export const editExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).set(updates)
      .then(() => {
        dispatch(editExpenseState(id, updates))
      })
      .catch((e) => {
        console.log('Error on edit expense', e)
      })
  }
}

export const setExpenses = (expenses) => ({
  type: actions.SET_EXPENSES,
  expenses
})

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once('value')
      .then((snapshot) => {
        const expenses = []
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setExpenses(expenses))
      })
  }
}
