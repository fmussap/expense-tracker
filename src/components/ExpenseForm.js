import React, { Component } from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

class ExpenseForm extends Component {
  constructor (props) {
    super()
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? props.expense.amount : 0,
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onAmountChange = (e) => {
    let amount = e.target.value
    if ( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      if (!amount) {
        amount = 0
      }
      this.setState(() => ({ amount }))
    }
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({ note }))
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }

  handleSubmit = (e) => {
    e.preventDefault()
    let error = ''
    if (!this.state.description) {
      error = 'Description field is required'
      this.setState(() => ({ error }))
    } else if (!this.state.amount) {
      error = 'Amount field is required'
      this.setState(() => ({ error }))
    } else {
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount),
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf()
      })
    }
    
  }

  render () {
    return (
      <form onSubmit={this.handleSubmit} className='form'>
        {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input
          className='text-input'
          type='text'
          placeholder='Description'
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          className='text-input'
          type='number'
          placeholder='Amount'
          value={this.state.amount}
          min="0" step="0.01"
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className='textarea'
          placeholder='Add a note for your expense'
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
          <button className='button'
            type='submit'>{this.props.expense ? 'Save Expense' : 'Add Expense'}
          </button>
        </div>
      </form>
    )
  }
}

export default ExpenseForm
