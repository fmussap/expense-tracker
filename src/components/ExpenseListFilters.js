import React, { Component } from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters'

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  }

  handleInput = (e) => (
    this.props.setTextFilter(e.target.value)
  )

  handleSelect = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate()
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount()
    }
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }

  render () {
    return (
      <div>
        <input type='text' value={this.props.filters.text} onChange={this.handleInput} />
        <select value={this.props.filters.sortBy} onChange={this.handleSelect}>
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          startDateId='start'
          endDateId='end'
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    )
  }
}

const mapStateToProps = ({ filters }) => ({
  filters
})

export default connect(mapStateToProps, {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
})(ExpenseListFilters)
