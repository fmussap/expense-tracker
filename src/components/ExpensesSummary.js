import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'

import SelectExpenses from '../selectors/expenses-selector'
import ExpensesTotal from '../selectors/expenses-total'

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  return (
    <div>
      <h1>
        Viewing {expenseCount} totalling {expensesTotal}
      </h1>
    </div>
  )
}

const mapStateToProps = ({ expenses, filters }) => {
  const visibleExpenses = SelectExpenses(expenses, filters)
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: numeral(ExpensesTotal(visibleExpenses)).format('$0,0.00')
  }
}

export default connect(mapStateToProps, undefined)(ExpensesSummary)
