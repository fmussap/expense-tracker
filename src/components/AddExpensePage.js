import React from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

export const AddExpensePage = (props) => {
  const handleSubmit = (expense) => {
    props.addExpense(expense)
    props.history.push('/')
  }
  return (
    <div>
      <div className='page-header'>
        <div className='content-container'>
          <h1>Add Expense</h1>
        </div>
      </div>
      <div className='content-container'>
        <ExpenseForm onSubmit={handleSubmit} />
      </div>
    </div>
  )
}

export default connect(undefined, {
  addExpense
})(AddExpensePage)
