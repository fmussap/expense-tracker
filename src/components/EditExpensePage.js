import React from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses'

export const EditExpensePage = (props) => {
  const handleSubmit = (expense) => {
    props.editExpense(props.expense.id, expense)
    props.history.push('/')
  }
  const handleClick = (e) => {
    props.removeExpense({ id: props.expense.id })
    props.history.push('/')
  }
  return (
    <div>
      <div className='page-header'>
        <div className='content-container'>
          <h1>Edit Expense</h1>
        </div>
      </div>
      <div className='content-container'>
        <ExpenseForm onSubmit={handleSubmit} expense={props.expense} />
        <button className='button button--danger' onClick={handleClick}>Remove expense</button>
      </div>
    </div>
  )
}

const mapStateToProps = ({ expenses }, props) => ({
  expense: expenses.find((expense) => expense.id === props.match.params.id)
})

export default connect(mapStateToProps, {
  editExpense,
  removeExpense
})(EditExpensePage)
