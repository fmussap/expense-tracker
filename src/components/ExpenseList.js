import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'
import ExpensesSummary from './ExpensesSummary'
import SelectExpenses from '../selectors/expenses-selector'

export const ExpenseList = (props) => {
  const renderList = () => {
    if (props.expenses.length === 0) {
      return <p>no expenses</p>
    }
    return props.expenses.map((expense) => {
      return <ExpenseListItem key={expense.id} {...expense} />
    })
  }
  return (
    <div>
      <ExpensesSummary />
      <h1>Expense List</h1>
      {renderList()}
    </div>
  )
}

const mapStateToProps = ({ expenses, filters }) => ({
  expenses: SelectExpenses(expenses, filters)
})

export default connect(mapStateToProps, undefined)(ExpenseList)
