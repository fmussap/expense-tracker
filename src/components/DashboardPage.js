import React from 'react'
import { Link } from 'react-router-dom'

import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'

const DashboardPage = () => (
  <div>
    <Link to='/create'>
      create
    </Link>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)

export default DashboardPage
