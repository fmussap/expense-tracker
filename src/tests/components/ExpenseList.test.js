import React from 'react'
import { shallow } from 'enzyme'

import { ExpenseList } from '../../components/ExpenseList'
import expenses from '../fixtures/expenses'

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render expenseListItem from each expense', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />)
  expect(wrapper.find('ExpenseListItem').length).toBe(expenses.length)
})

test('should not render expenseListItem when there is no expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />)
  expect(wrapper.find('ExpenseListItem').length).toBe(0)
})

test('should show "no expenses" when there is no expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />)
  expect(wrapper.find('p').text()).toBe('no expenses')
})
