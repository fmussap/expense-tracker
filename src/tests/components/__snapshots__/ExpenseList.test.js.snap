// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <Connect(ExpensesSummary) />
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={5}
    createdAt={1514282568444}
    description="banana"
    id={1}
    key="1"
    note="note test"
  />
  <ExpenseListItem
    amount={8}
    createdAt={1515455368444}
    description="apple"
    id={2}
    key="2"
    note="note test 2"
  />
</div>
`;
